ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java

```diff
@@ -106,7 +107,11 @@ public JsonRpcResponse response(final JsonRpcRequestContext request) {
             .map(blockHeader -> blockHeader.getBaseFee().orElse(Wei.ZERO))
             .orElseGet(
                 () ->
-                    Optional.of(protocolSchedule.getByBlockNumber(nextBlockNumber).getFeeMarket())
+                    Optional.of(
+                            protocolSchedule
+                                .getForNextBlockHeader(
+                                    chainHeadHeader, chainHeadHeader.getTimestamp())
```

Comment:
Think we discussed having a comment here to explain why we're using head timestamp instead of timestamp for next block?

Reply:
We can only estimate the timestamp for the next block by adding the seconds per slot to the head block's timestamp i.e. add 12 when it's mainnet. This is only an estimate though as it's possible that a block could be skipped due to a validator not proposing a block.

I'm using the parent headers timestamp which is wrong but is simple. However, since the fee market doesn't change in Shanghai or Cancun it's adding a new fee market not changing the base fee one we can get away with this simple implementation for now.

The better approach we considered was to check the difference between timestamps. There are many edge cases with this though. We need to consider the possibility of the blocks that are missed. And what to do if we don't have all the block headers.

Perhaps we should just add some predefined constants for the networks though?

Reply:
Unfortunately, there's no way of accurately predicting the next block's timestamp, which is a significant difference between block number vs timestamp for switching fork.

e.g. we have a network with 10 slots per epoch, 1 second per slot, genesis @ t = 1 (Epoch 1, Slot 1).
We have a fork configured at Epoch 2, Slot 1 @ t = 11 (forks are currently always configured at the start of an Epoch).
Let's say the proposers miss Epoch 1, Slot 9, 10 and Epoch 2, Slot 1, 2, 3
Last block of previous fork is Epoch 1, Slot 8 @ t = 8; First block of next fork is Epoch 2, Slot 4 @ t = 14.

So Block 8 @ t = 8 -> (fork @ t = 11) -> Block 9 @ t = 14

Even if we know slots per epoch = 1, then we will still predict that Block 9 should be @ t = 9 which won't hit the fork.

A dummy value of 1 minute might be pragmatic and future-proof but feels very hacky 🙈

Reply:
You'd have to fix both sides of the computation of this I think, which would look something like
 - block timestamp EIP to suggest that block timestamp is always the slot -> timestamp computation, so the EL block timestamp is locked to the slot (predictably computed as expected slot time)
 - then your computation here is trivial, but currently you've theoretically got a 12 sec margin of error... we're asking for a block 4 seconds into the slot before you're computing a block for, and the block could be up to 4 seconds late (then it'd be reorged out if a CL spec change suggestion gets up).
 
 If all EL's could converge on using the timestamp of the block you're proposing for, then this is a trivial thing, but I think ideally it'd be in spec (I haven't checked if it's already specified in that way but i'd be surprised if it is...)

Reply:
@fab-10 I'm guessing the reason you need it is related to an optimisation rather than part of the specification itself? I am assuming that if we're querying the txpool then we are building a block and therefore should have a timestamp from the CL.

Reply:
In this specific case for eth_feeHistory since the fee market is either pre-london or post-london we don't need to worry about an exact timestamp. But if the fee market is changed in a future fork then I think we will need to have something in a spec.

Reply:
I wanted to avoid adding general a getForNextBlockHeader method that doesn't take a timestamp as I don't think there is a simple answer that isn't a hack right now and might depend on the context. The safest way is to always provide the timestamp which we do have when we are proposing a block.

@fab-10 I will leave this open for another day for you to have another look and add comments or block it if you don't think it shouldn't go in as is. I think this is safe for the eth_feeHistory case but we need to come up with a better way for the general case and that might require some spec changes as well.

Reply:
The part I am reviewing is not the block production, but when accepting transactions to the pool, it is an edge case that could only affect fee computation just before a hard fork, but after the hard fork the computation will be fine again, so I am fine with this PR since the only side effect seems to be a non optimal block production in case we are the ones that build the hard fork block.